### ISSUE TITLE
Enhancing model caching to automatically reload model when they changed (e.g., on disc)

### ISSUE THREAD

**Issue by goto40** [textX/textX#199]
I would like to enhance model caching to allow to automatically reload models (e.g., from disc) when models changed (e.g., on disc).

Need further analysis to understand how to allow a seamless integration into the (scope provider) logic of different predicates. Possible predicates:
 - date of file changed
 - MD5 of file changed
 - in-memory representation within an editor changed (?)

Also allow to update models.

**Example use case / unittest** with global repository:
 - load A,B,C (A is including B and C)
 - reload A (B,C are cached)
 - change B on disc
 - reload A (C is cached, B is reloaded)

**Example use case / unittest** with global repository:
 - load A,B,C (A is including B and C) + validate A,B,C
 - reload A (B,C are cached) + validate A (B,C are not validated because they are cached)
 - update B with update mechanism + validate B; --> also reload (request reload?) of all model including B

**Comment by igordejanovic** [comment]
Good idea but not sure where this should fit. Should it be done in the core textX or should we do that in textX-LS? We should definitely have some generic extension mechanism in the model loading/caching process that could be used to implement this. There are many ways we could detect model changes. By pooling date/MD5 as you suggested or by using some platform specific feature like [inotify](https://en.wikipedia.org/wiki/Inotify) for example.

**Comment by goto40** [comment]
I had the idea to provide some callback mechanism to hook-in the scope provider logic caching files. This hook could be implemented by textx-ls.

Let's think about it...

**Comment by goto40** [comment]
If you have ideas, we can collect requirements / use cases here... (for the new mechanism).

**Comment by danixeee** [comment]
I have already started implementing this in `textX-LS` (not yet completed).
When textX language project is installed in _editable_ mode, file system watcher is registered for its project path. When `.tx` files are changed extension is regenerated and updated (see [this](https://github.com/textX/textX-LS/blob/danixeee/textx-view-container/client/src/services/projectService.ts#L160)). I need to add watcher for `.py` files as well, and reload meta-models for already opened documents (no need to update extension, because coloring depends just on grammar keywords).

Thoughts on that approach?

**Comment by goto40** [comment]
Hm. Not sure.

If we can handle it completely within textX-ls, perfect. Maybe we need support from the textx core api.

Check the use cases above. We should define what should happen when files and included files are modified. When to reload files...

I think we may need an api call to invalidate cached files... Maybe the new `get_included_files` api function is enough?

**Comment by goto40** [comment]
However, I think we should focus on model files (e.g., `*.eflow`) and not meta model files (`*.tx`, `*.py`).

The end user wants to use the dsl, not to develop the dsl...

**Comment by igordejanovic** [comment]
Actually, we have two classes of users, DSL designers and DSL users. We are targeting both with our tools.

**Comment by igordejanovic** [comment]
Maybe we need a call in textX to properly invalidate cache based on the path that has been changed. In that way we leave to the user of the library (in this case textX-LS) to implement mechanism to detect which file has been changed.
